File: fabric/src/main/java/dev/enchantaholic/custom/MidasTouch.java
package dev.enchantaholic.custom;

import dev.enchantaholic.core.CustomMath;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.level.block.Block;

/** Midas Touch: 3×L % chance per break to drop a gold nugget; from level 34 that drop is sometimes an ingot. */
public final class MidasTouch {
	private MidasTouch() {}

	static void onBlockBreak(ServerLevel level, ServerPlayer player, BlockPos pos) {
		int lvl = CustomEnchants.mainHand(player, CustomEnchants.MIDAS_TOUCH);
		if (!CustomMath.roll(CustomMath.midasNuggetPercent(lvl), level.getRandom().nextInt(100))) return;
		boolean ingot = CustomMath.roll(CustomMath.midasIngotPercent(lvl), level.getRandom().nextInt(100));
		Block.popResource(level, pos, new ItemStack(ingot ? Items.GOLD_INGOT : Items.GOLD_NUGGET));
	}
}

File: bedrock/src/adapters/custom/midas-touch.ts
import { ItemStack, type PlayerBreakBlockAfterEvent } from "@minecraft/server";
import { midasDrop } from "../../core/custom/math";
import { CUSTOM } from "../../core/custom/roster";
import { defaultRng, type Rng } from "../../core/rng";
import { getCustomLevel } from "../item-levels";
import { activePlayer, isCreative } from "./gate";

const DROPS = { nugget: "minecraft:gold_nugget", ingot: "minecraft:gold_ingot" } as const;

/** Midas Touch: 3 % per level to drop a gold nugget (an ingot sometimes at level 34+). Not in creative. */
export function midasTouch(ev: PlayerBreakBlockAfterEvent, rng: Rng = defaultRng): void {
  const player = ev.player;
  if (!activePlayer(player) || isCreative(player)) return;
  const level = getCustomLevel(ev.itemStackBeforeBreak, CUSTOM.midas_touch);
  if (level <= 0) return;
  const drop = midasDrop(level, rng);
  if (!drop) return;
  ev.dimension.spawnItem(new ItemStack(DROPS[drop], 1), ev.block.center());
}
